### PR TITLE
azure-python-devtools port

### PR DESCRIPTION
**Pull request by v-iam**
This pull request introduces a dependency on [azure-python-devtools](https://github.com/Azure/azure-python-devtools), which contains non-CLI-specific code extracted from azure-cli-testsdk so that it could be shared by both the CLI and [Azure SDK for Python](https://github.com/Azure/azure-sdk-for-python) test code.

The azure-cli-testsdk directory now serves as a compatibility layer between the CLI tests and azure-python-devtools; it contains the CLI-specific modules and configuration which used to be static but is now exposed (e.g. the multiple kwargs to ScenarioTest).

Tests themselves should require minimal or no modification to work in the new setup; the storage tests all ran successfully without modification.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,6 +1,7 @@
 adal==0.4.3
 applicationinsights==0.10.0
 argcomplete==1.8.0
+azure_devtools==0.1.0
```
Comment (troydai) [inline requirements.txt]: underline in package name doesn't seem to be the norm of python packages.

Comment (troydai) [inline requirements.txt]: And I don't think it needs to be listed in requirements list. In dependencies is good enough.

File: src/azure-cli-testsdk/azure/cli/testsdk/__init__.py [hunk 2]
```diff
@@ -15,6 +15,6 @@
 __all__ = ['ScenarioTest', 'LiveTest', 'ResourceGroupPreparer', 'StorageAccountPreparer',
            'RoleBasedServicePrincipalPreparer', 'CliTestError', 'JMESPathCheck', 'JMESPathCheckExists', 'NoneCheck',
            'live_only', 'record_only', 'StringCheck', 'StringContainCheck', 'get_sha1_hash', 'KeyVaultPreparer',
-           'JMESPathCheckGreaterThan', 'api_version_constraint']
+           'JMESPathCheckGreaterThan', 'api_version_constraint', 'execute']
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/__init__.py]: Why do you export `execute` method?

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py [hunk 3]
```diff
@@ -3,29 +3,29 @@
...
+
+from azure_devtools.scenario_tests.base import LiveTest
+from azure_devtools.scenario_tests.base import ScenarioTest as DevToolsScenarioTest
+from azure_devtools.scenario_tests.recording_processors import (
+    SubscriptionRecordingProcessor, OAuthRequestResponsesFilter,
+    GeneralNameReplacer, LargeRequestBodyProcessor,
+    LargeResponseBodyProcessor, LargeResponseBodyReplacer,
+    DeploymentNameReplacer
+)
+from azure_devtools.scenario_tests.utilities import create_random_name
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: I prefer all these names are exposed in one namespace instead of scattering in various namespaces, which hurts discoverability.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py [hunk 4]
```diff
@@ -3,29 +3,29 @@
...
-import tempfile
-import shutil
 import logging
-import six
-import vcr
+import inspect
+import sys
+
+from azure_devtools.scenario_tests.base import LiveTest
+from azure_devtools.scenario_tests.base import ScenarioTest as DevToolsScenarioTest
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: Can we have a batter name than `DevToolsScenarioTest`? If it is a base class, called it `ScenarioBasedTest`.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py [hunk 5]
```diff
@@ -46,128 +46,49 @@ def find_recording_dir(test_file):
...
+                patch_long_run_operation_delay,
+                patch_load_cached_subscriptions,
+                patch_retrieve_token_for_user,
+            ]
+        )
+        kwargs.setdefault(
+            'recording_dir',
+            find_recording_dir(inspect.getfile(self.__class__))
+        )
+        super(ScenarioTest, self).__init__(method_name, **kwargs)
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: Please avoid using `kwargs`. It leads to unreadable code.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py [hunk 6]
```diff
@@ -297,7 +163,9 @@ def _in_process_execute(self, command):
         try:
             # issue: stderr cannot be redirect in this form, as a result some failure information
             # is lost when command fails.
-            self.exit_code = cli_main(shlex.split(command), file=output_buffer) or 0
+            cmd_parts = [part.replace('"', '')
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: Be careful of this replacing logic it converts `"\""` to `\`. What you actually try to do is `strip` the double quotes.

File: src/azure-cli-testsdk/azure/cli/testsdk/decorators.py [hunk 7]
```diff
@@ -1,24 +1,8 @@
-# --------------------------------------------------------------------------------------------
-# Copyright (c) Microsoft Corporation. All rights reserved.
-# Licensed under the MIT License. See License.txt in the project root for license information.
-# --------------------------------------------------------------------------------------------
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/decorators.py]: Do not remove file header.

File: src/azure-cli-testsdk/azure/cli/testsdk/decorators.py [hunk 8]
```diff
@@ -1,24 +1,8 @@
...
 import unittest
-from azure.cli.core.profiles import supported_api_version
-from .const import ENV_LIVE_TEST
-
 
-def live_only():
-    return unittest.skipUnless(
-        os.environ.get(ENV_LIVE_TEST, False),
-        'This is a live only test. A live test will bypass all vcrpy components.')
+from azure_devtools.scenario_tests.decorators import live_only, record_only
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/decorators.py]: It is unnecessary to import them here. Import and export them in `__init__.py`

File: src/azure-cli-testsdk/azure/cli/testsdk/patches.py [hunk 9]
```diff
@@ -3,6 +3,8 @@
 # Licensed under the MIT License. See License.txt in the project root for license information.
 # --------------------------------------------------------------------------------------------
 
+from azure_devtools.scenario_tests.patches import patch_time_sleep_api, _mock_in_unit_test
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/patches.py]: same as above.

File: src/azure-cli-testsdk/azure/cli/testsdk/preparers.py [hunk 10]
```diff
@@ -3,108 +3,15 @@
...
 
-import inspect
-import functools
 import os
 
-from .base import ScenarioTest, execute
+from azure_devtools.scenario_tests.base import ScenarioTest
+from azure_devtools.scenario_tests.utilities import create_random_name
+from azure_devtools.scenario_tests.recording_processors import RecordingProcessor
+from azure_devtools.scenario_tests.preparers import AbstractPreparer, SingleValueReplacer
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/preparers.py]: Move to `__init__.py`

File: src/azure-cli-testsdk/azure/cli/testsdk/recording_processors.py [hunk 11]
```diff
@@ -1,156 +1 @@
...
-    def process_response(self, response):
-        for old, new in self.names_name:
-            if response['body']['string']:
-                response['body']['string'] = response['body']['string'].replace(old, new)
-
-            self.replace_header(response, 'location', old, new)
-            self.replace_header(response, 'azure-asyncoperation', old, new)
-
-        return response
+from azure_devtools.scenario_tests.recording_processors import *
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/recording_processors.py]: Same as above,

File: src/azure-cli-testsdk/azure/cli/testsdk/utilities.py [hunk 12]
```diff
@@ -1,36 +1 @@
...
-    sha1 = hashlib.sha256()
-    with open(file_path, 'rb') as f:
-        while True:
-            data = f.read(65536)
-            if not data:
-                break
-            sha1.update(data)
-
-    return sha1.hexdigest()
+from azure_devtools.scenario_tests.utilities import create_random_name, get_sha1_hash
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/utilities.py]: Same as above.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 4 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: Is `ScenarioTestBase` OK? I feel like I've seen that naming convention more often.

**Comment by troydai** [comment]
@v-iam @lmazuel I need to review the code in https://github.com/Azure/azure-python-devtools/commits/master first before we can adopt this change. Is the package there already stabilized or it is still being developed.

**Comment by v-iam** [comment]
azure-python-devtools has a 0.1.0 release which you can review. The only development that will happen is to respond to your review and any changes that need to be made to make Travis happy.

**Comment by troydai** [comment]
Ok. Cool. So please take a look at my comments in this PR first. I'll take a deeper dive into devtools repository.

File: src/azure-cli-testsdk/azure/cli/testsdk/__init__.py (hunk 2 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/__init__.py]: Hm, at some point maybe I got the impression that something used it, but I think I was mistaken. I'll remove that.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 3 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: Do you mean you want me to pull them all into the top level of `azure_devtools.scenario_tests` via `__init__.__all__`?

File: src/azure-cli-testsdk/azure/cli/testsdk/patches.py (hunk 9 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/patches.py]: `_mock_in_unit_test` shouldn't be exported in `__init__.py`, though, should it?

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 4 above)
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: Is the base class an abstract class (technically no such concept in Python).

Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: No, I believe it should be usable as is.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 3 above)
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: No. `from xxx import *` is a bad practice and no one should do. I want to simply the namespace. I do not want the users to get components from different namespaces because it is not easy to figure out. All names should be exposed in the same package. https://github.com/Azure/azure-cli/blob/master/src/azure-cli-testsdk/azure/cli/testsdk/__init__.py

File: src/azure-cli-testsdk/azure/cli/testsdk/patches.py (hunk 9 above)
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/patches.py]: If it doesn't need to be exposed and it is not used here, why do you import it?

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 3 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: Yes, that's what I meant. I wasn't proposing using an `import *`.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 4 above)
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: `ReplayableTest` sounds better.

File: src/azure-cli-testsdk/azure/cli/testsdk/patches.py (hunk 9 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/patches.py]: It is used here; almost every patch uses it. But my impression is that it's an internal helper for patches, not something that consumers of azure-cli-testsdk should be using, and thus shouldn't be exposed.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 4 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: OK, works for me. Do you want me to change the name in `azure-devtools` itself, or just the name of the import here?

File: requirements.txt (hunk 1 above)
Comment (lmazuel) [inline requirements.txt]: - Never use an underscore in published package name ([see this pip issue for instance](https://github.com/pypa/pip/issues/3212))
- I have no opinion on this file or `scripts/setup.py`, since I don't know where the later is actually called

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 4 above)
Comment (lmazuel) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: You can change the name in the initial package for clarity

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 3 above)
Comment (lmazuel) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: `__all__` LGTM

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 5 above)
Comment (lmazuel) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: Agreed, signature should be:
```python
def __init__(self, method_name, recording_processors, replay_processors, recording_patches, replay_patches, recording_dir):
```
but this makes a lot of parameters :/

I was wondering if at this point we should just not take a "vcr" parameter:
```
        vcr = vcr.VCR(
            cassette_library_dir=recording_dir,
            before_record_request=self._process_request_recording,
            before_record_response=self._process_response_recording,
            decode_compressed_response=True,
            record_mode=self.config.record_mode,
            filter_headers=self.FILTER_HEADERS
        )
```
and then
```python
def __init__(self, method_name, vcr, recording_patches, replay_patches):
```

This removes the need for "recording_dir", "recording_processors" and "replay_processors" and even simplify the SDK about "record_mode".

Thoughts @v-iam @troydai ?

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py [hunk 13]
```diff
@@ -177,68 +98,13 @@ def create_random_name(self, prefix, length):
...
-            for processor in self.recording_processors:
-                request = processor.process_request(request)
-                if not request:
-                    break
         else:
-            for processor in self.replay_processors:
-                request = processor.process_request(request)
-                if not request:
-                    break
+            return moniker
```
Comment (lmazuel) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: Remove this `else`, try to respect as close as possible the initial code if this is not an improvement but just detail, to simplify review.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 6 above)
Comment (lmazuel) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: Why did you change that? I mean, this is not the point of the port right? PR should be atomic and this one is about "port to azure_devtools". If this is an improvement this is great, but should be in a separate PR. This PR is important enough to avoid this kind of details to be shipped at the same time IMHO.

File: src/azure-cli-testsdk/azure/cli/testsdk/decorators.py (hunk 7 above)
Comment (lmazuel) [inline src/azure-cli-testsdk/azure/cli/testsdk/decorators.py]: Agreed

File: src/azure-cli-testsdk/azure/cli/testsdk/decorators.py (hunk 8 above)
Comment (lmazuel) [inline src/azure-cli-testsdk/azure/cli/testsdk/decorators.py]: The point here was to keep compatibilty of people doing `from azure.cli.testsdk.decorators import live_only`. If this is not required, we can remove completely this line.

File: src/azure-cli-testsdk/azure/cli/testsdk/preparers.py [hunk 14]
```diff
@@ -3,108 +3,15 @@
 # Licensed under the MIT License. See License.txt in the project root for license information.
 # --------------------------------------------------------------------------------------------
 
-import inspect
-import functools
 import os
 
-from .base import ScenarioTest, execute
+from azure_devtools.scenario_tests.base import ScenarioTest
```
Comment (lmazuel) [inline src/azure-cli-testsdk/azure/cli/testsdk/preparers.py]: Conceptually, you are not importing the same ScenarioTest. I think this should still be `from .base import ScenarioTest` to import the one from CLI, and not the base one (unless you are 100% sure the base class is enough here).

File: src/azure-cli-testsdk/azure/cli/testsdk/preparers.py (hunk 10 above)
Comment (lmazuel) [inline src/azure-cli-testsdk/azure/cli/testsdk/preparers.py]: Same here, the point was to keep compat with `from azure.cli.testsdk.preparers import AbstractPreparer`

File: src/azure-cli-testsdk/azure/cli/testsdk/recording_processors.py (hunk 11 above)
Comment (lmazuel) [inline src/azure-cli-testsdk/azure/cli/testsdk/recording_processors.py]: Same, compat only. If you think this is internal enough, the whole file can disappear.

**Comment by lmazuel** [comment]
@v-iam note that it's probably the underscore that kills Travis and pip

File: requirements.txt (hunk 1 above)
Comment (v-iam) [inline requirements.txt]: The published package name uses a hyphen: https://pypi.python.org/pypi/azure-devtools

The underscore was erroneous, and is in any case now removed entirely from `requirements.txt` per @troydai's comment.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 6 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: I don't mind removing it, but I made it because it was preventing me from running the tests locally. I'll revert it for this PR.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 5 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: Thoughts: it could work. The upside (simplified `__init__` signature) is obvious, so here are the downsides I see:
* Introduces a direct dependency on `vcrpy` for anything that uses `ReplayableTest`. As it is now, only `azure-devtools` depends on `vcrpy`.
* If I understand correctly, this would require subclasses like `ScenarioTest` to override `_process_request_recording` and `_process_response_recording` to add recording/replay processors. Is this actually better than having a lot of kwargs? It's certainly more difficult.

File: src/azure-cli-testsdk/azure/cli/testsdk/preparers.py (hunk 14 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/preparers.py]: Actually I don't think it's needed at all.

File: src/azure-cli-testsdk/azure/cli/testsdk/recording_processors.py (hunk 11 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/recording_processors.py]: In the absence of someone firmly telling me otherwise I'm going to take these "import-through"s back out. Leaving them in was for the sake of caution but if @troydai wants them out we can probably assume it's safe.

File: scripts/setup.py [hunk 15]
```diff
@@ -25,6 +25,7 @@
 
 DEPENDENCIES = [
     'autopep8>=1.2.4',
+    'azure_devtools>=0.2.0',
```
Comment (v-iam) [inline scripts/setup.py]: Oops.

File: src/azure-cli-testsdk/azure/cli/testsdk/utilities.py (hunk 12 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/utilities.py]: `test_rdbms_commands` imports `create_random_name` from `azure.cli.testsdk.utilities` directly; removing this causes that import to fail. Should I leave this import or expose `create_random_name` in `__init__.py` and change the test to import from there instead?

**Comment by troydai** [comment]
Please fix the CI failures.

**Comment by v-iam** [comment]
I'm working on it., but I'm unable to repro them. I'm going to see if I can do so using the Docker container.

**Comment by v-iam** [comment]
@troydai @lmazuel CI failures fixed. Let me know what if anything else still needs to be done here.

**Comment by v-iam** [comment]
[This question](https://github.com/Azure/azure-cli/pull/3621/files/a77f118633223060f34c653065deade5c4076628#r120485476) still needs an answer. As it stands right now, `create_random_name` and `get_sha1_hash` are duplicated between azure-cli and azure-devtools.

**Comment by lmazuel** [comment]
@troydai @johanste about the question @v-iam mentions, the idea is to have this kind of "empty" file
```python
from somewhere_i_extracted_the_code import my_method
```
to be sure we can still import `my_method` exactly like before.

To decide if we push people to use the new import (meaning: ask them to migrate) or if we keep this "empty" file is up to you.

**Comment by troydai** [comment]
@v-iam @lmazuel we need to clear the CLA-required labl.

File: src/azure-cli-testsdk/azure/cli/testsdk/patches.py (hunk 9 above)
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/patches.py]: If it is indeed needed and exposed from devtools, it should be renamed since it is not private anymore.

File: src/azure-cli-testsdk/azure/cli/testsdk/decorators.py (hunk 8 above)
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/decorators.py]: Tests should not import symbols from lower level namespace. Let's force all tests to import names from top level.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 13 above)
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: I think else-return will trigger pylint error.

File: src/azure-cli-testsdk/azure/cli/testsdk/__init__.py [hunk 16]
```diff
@@ -3,14 +3,16 @@
 # Licensed under the MIT License. See License.txt in the project root for license information.
 # --------------------------------------------------------------------------------------------
 
-from .base import ScenarioTest, LiveTest
+from azure_devtools.scenario_tests import (LiveTest, live_only, record_only, get_sha1_hash)
```
Comment (troydai) [inline src/azure-cli-testsdk/azure/cli/testsdk/__init__.py]: The parentheses are unnecessary.

File: src/azure-cli-testsdk/azure/cli/testsdk/base.py (hunk 13 above)
Comment (v-iam) [inline src/azure-cli-testsdk/azure/cli/testsdk/base.py]: The `else` is gone.

**Comment by lmazuel** [comment]
@troydai The bot already said "You don't need to sign CLA", but put a "CLA required" label. Seems like a bot issue no?

**Comment by v-iam** [comment]
I had thought maybe I accidentally included some commits from my personal GitHub account when I worked from home, but there are no commits under that account included in this PR. So I don't know what's needed for the CLA.

**Comment by troydai** [comment]
I'll take another look today.

**Comment by troydai** [comment]
@v-iam looks good to me now. Please rebase it on the master. As soon as the CI passes, I'll merge it.

**Comment by v-iam** [comment]
OK, I think maybe I did the wrong thing here.

**Comment by troydai** [comment]
@v-iam there are two files without license headers which failed the CI.

**Comment by v-iam** [comment]
Yup, just saw that, possibly lost in merges. Fixing.

**Comment by v-iam** [comment]
CI looks good to go @troydai

**Comment by troydai** [comment]
I notice the code coverage went down fairly dramatically. Did we miss any tests?

**Comment by v-iam** [comment]
Sorry about that. I think it is probably because of the issue fixed in this PR: https://github.com/Azure/azure-python-devtools/pull/8

So we should release a new devtools with that fix in and then update to use that.

**Comment by troydai** [comment]
Ok. That make sense. Can you rerun the tests and update the coverage report? I'd like to make sure tests are not skipped. (You can merge the commit and force push to trigger a new coverage report).

**Comment by v-iam** [comment]
Will do. Thanks for merging that other PR.

**Comment by v-iam** [comment]
OK, if I understand correctly, I think the coverage is better now: https://codecov.io/gh/Azure/azure-cli/pull/3621/changes

**Comment by troydai** [comment]
Ok. That looks good.
